book.md

```diff
@@ -337,7 +337,153 @@ purposes of this chapter we will focus on:
+
+[^build-options]: The code is generated with the `-fno-stack-protector` option,
+  to ensure GCC's stack guard feature is disabled. We also used the `-O1`
+  optimization level.
+
+The exact details of the stack frame layout, including the ordering of
+variables and the exact control information stored, will depend on the specific
+compiler version we're using and the architecture we're compiling for.
+
+As can be seen the stack diagram, an overflowing write can overwrite the saved
+frame pointer (FP) and link register (LR), which contains the return address,
+in `main`'s frame. This is an exploitable vulnerability.
+
+Before non-executable stacks were mainstream, a common way to exploit these
+vulnerabilities would be to use the overflow to simultaneously write
+shellcode[^shellcode]\index{shellcode} to the stack and overwrite the return
+address so that it points to the shellcode. [@AlephOne1996] is a classic
+example of this technique.
+
+[^shellcode]: A shellcode is a short instruction sequence that performs an
+  action such as starting a shell on the victim machine.
+
+The obvious solution to this issue is to use memory protection features of the
+processor in order to mark the stack (along with other data sections) as
+non-executable[^trampolines]. However, even when the stack is not executable,
+more advanced techniques can be used to exploit an overflow that overwrites the
+return address. These take advantage of code that already exists in the
+executable or in library code, and will be described in the next section.
+
+[^trampolines]: Note that the use of [nested
+  functions](https://gcc.gnu.org/onlinedocs/gcc/Nested-Functions.html) in GCC
+  requires [trampolines](https://gcc.gnu.org/onlinedocs/gccint/Trampolines.html)
+  which reside on an executable stack. The use of nested functions, therefore,
+  poses a security risk.
+
+Stack canaries are an alternative mitigation for stack buffer overflows. The
+general idea is to store a known value, called the stack canary, between the
+buffer and the control information (in the example, the saved FP and LR), and
+to check this value before leaving the function. Since an overflow that would
+overwrite control data is going to overwrite the canary first, a corruption of
+the control data through a stack buffer overflow will be detected.
+
+This technique has a few limitations: first of all, it specifically aims to
+protect against stack buffer overflows, and does nothing to protect against
+stronger primitives (e.g. arbitrary write primitives). Control-flow integrity
+techniques, which are described in the next section, aim to protect the
+integrity of control data against any modification.
+
+Since stack canaries come with additional checks, compilers usually employ
+heuristics to determine which functions are considered vulnerable and therefore
+include the mitigation in their generated code. Since heuristics aren't always
```

Comment:
small nit/suggestion:

> ...heuristics to determine which functions are considered vulnerable and only include the mitigation in the generated code of those functions.

In the original wording, I think 'their' could be referring to the compiler's generated code (all of it) or the vulnerable functions' generated code, though I might just be overthinking it though.

Reply:
I think you are right. I'll rephrase this slightly to make this clearer, thanks!